Add support for client assertion signing key authentication

kaczowkad opened the pull request:
### Changes

The `omniauth-auth0` gem only supports client secret authentication.  [Auth0 documentation](https://auth0.com/docs/secure/application-credentials#client-secret-authentication) states: 

```
To ​​improve your security posture, we recommend using the Private Key JWT authentication method.
```

Opened issue #199 on 2024-11-07 to inquire about adding support for private key JWT authentication.  We created a workaround to support private key JWT authentication in our application using the `omniauth-auth0` gem.  Opening this PR to get feedback on adding support for private key JWT added to the gem.  

This PR adds two optional parameters to the config: 
 - client_assertion_signing_key - the private key
 - client_assertion_signing_algorithm - algorithm defaults to RS256 

The current interface is preserved. Client secret authentication can be used without changing the interface. 

Client secret authentication: 
```
AUTH0_CONFIG = Rails.application.config_for(:auth0)

Rails.application.config.middleware.use OmniAuth::Builder do
  provider(
    :auth0,
    AUTH0_CONFIG['auth0_client_id'],
    AUTH0_CONFIG['auth0_client_secret'],
    AUTH0_CONFIG['auth0_domain'],
    callback_path: '/auth/auth0/callback',
    authorize_params: {
      scope: 'openid profile'
    }
  )
end
```

Client assertion signing key: 
```
AUTH0_CONFIG = Rails.application.config_for(:auth0)

Rails.application.config.middleware.use OmniAuth::Builder do
  provider(
    :auth0,
    AUTH0_CONFIG['auth0_client_id'],
    nil,
    AUTH0_CONFIG['auth0_domain'],
    callback_path: '/auth/auth0/callback',
    authorize_params: {
      scope: 'openid profile'
    }
    client_assertion_signing_key: AUTH0_CONFIG[:auth0_client_assertion_signing_key],
    client_assertion_signing_algorithm: AUTH0_CONFIG[:auth0_client_assertion_signing_algorithm]}
  )
end
```

The existing code has `rubocop` violations which were not addressed because it was outside the scope of these changes.  No new `rubocop` violations were introduced.  

### References

Please include relevant links supporting this change such as a:

 - issue #199
 - [form post](https://community.auth0.com/t/ruby-on-rails-with-private-key-jwt/150744) https://community.auth0.com/t/ruby-on-rails-with-private-key-jwt/150744 which didn't address the `omniauth-auth0` gem


### Testing

Duplicated all the specs in `spec/omniauth/strategies/auth0_spec.rb` for client assertion signing key.  The existing specs were scoped to 'client secret authentication'.  All specs passed.  

Testing with Auth0 requires an application setup for private key JWT authentication.  The public key must be uploaded to the application in Auth0.  

* [x] This change adds unit test coverage
* [ ] This change has been tested on the latest version of the platform/language or why not
   Tested with ruby 3.2.2.  Built the gem and client assertion signing key authentication with a Ruby on Rails application.

### Checklist

* [x] I have read the [Auth0 contribution guidelines](https://github.com/auth0/open-source-template/blob/master/GENERAL-CONTRIBUTING.md)
* [x] I have read the [Auth0 Code of Conduct](https://github.com/auth0/open-source-template/blob/master/CODE-OF-CONDUCT.md)
* [x] All existing and new tests complete without errors
* [x] All code quality tools/guidelines in the [CONTRIBUTING documentation](https://github.com/auth0/omniauth-auth0/blob/master/CONTRIBUTING.md) have been run/followed

kaczowkad committed:
Add support for client assertion signing key authentication while continuing to support client secret authentication.
The client assertion signing key allows Private Key JWT application authentication in Auth0.  The client_assertion_signing_key and client_assertion_signing_algorithm are optional parameters.
Client assercition signing key will be used if a client_assertion_signing_key is privided. The client_assertion_signing_algorithm is optional. The algorithm defaults to RS256 if nil or not
provided.

kaczowkad committed:
Add support for client assertion signing key authentication while continuing to support client secret authentication.
The client assertion signing key allows Private Key JWT application authentication in Auth0.  The client_assertion_signing_key and client_assertion_signing_algorithm are optional parameters.
Client assercition signing key will be used if a client_assertion_signing_key is privided. The client_assertion_signing_algorithm is optional. The algorithm defaults to RS256 if nil or not
provided.

arpit-jn commented on README.md:
````diff
@@ -54,17 +54,34 @@ Adding the SDK to your Rails app requires a few steps:
+```
+**Note**: you must upload the corresponding public key to your Auth0 tenant, so that Auth0 is able to verify the JWT signature.
+
+client_assertion_siging_algorithm is optional and defaults to RS256.
````
nit: Typo: "siging" → "signing"

arpit-jn commented on README.md:
```diff
@@ -82,6 +99,27 @@ Rails.application.config.middleware.use OmniAuth::Builder do
+    callback_path: '/auth/auth0/callback',
+    authorize_params: {
+      scope: 'openid profile'
+    }
```
This code block has two syntax errors that will crash on boot:

1. Missing comma after `authorize_params` hash (line 115)
2. Stray `}` before closing `)` (line 117)

```ruby
# Current (broken)
    authorize_params: {
      scope: 'openid profile'
    }
    client_assertion_signing_key: AUTH0_CONFIG[:auth0_client_assertion_signing_key],
    client_assertion_signing_algorithm: AUTH0_CONFIG[:auth0_client_assertion_signing_algorithm]}
  )

# Should be
    authorize_params: {
      scope: 'openid profile'
    },
    client_assertion_signing_key: AUTH0_CONFIG[:auth0_client_assertion_signing_key],
    client_assertion_signing_algorithm: AUTH0_CONFIG[:auth0_client_assertion_signing_algorithm]
  )

arpit-jn commented on README.md:
```diff
@@ -54,17 +54,34 @@ Adding the SDK to your Rails app requires a few steps:
+development:
+  auth0_domain: <YOUR_DOMAIN>
+  auth0_client_id: <YOUR_CLIENT_ID>
+  auth0_client_assertion_signing_key: <YOUR AUTH0 CLIENT ASSERTION SIGNING PRIVATE KEY>
```
question: The YAML config will yield a PEM string, but JWT.encode expects an OpenSSL::PKey object. Users will get a cryptic TypeError at runtime.

Either handle the conversion in code:

`key = key.is_a?(String) ? OpenSSL::PKey::RSA.new(key) : key`

Or document the required conversion in the initializer example:

`client_assertion_signing_key: OpenSSL::PKey::RSA.new(AUTH0_CONFIG['auth0_client_assertion_signing_key'])`

kaczowkad replied:
@arpit-jn Do you have a preference on which way to handle this?  To handle the conversion in the code it would need to support EC keys as well, correct?

kaczowkad replied:
Documented the required conversion in the initializer.

arpit-jn replied:
Documentation approach works, good call. Since Auth0 only supports RSA-based algorithms (RS256, RS384, PS256) for Private Key JWT, you don't need to worry about EC key conversion here. The README example with OpenSSL::PKey::RSA.new(...) is appropriate.

arpit-jn commented on lib/omniauth/auth0/jwt_token.rb:
```diff
@@ -0,0 +1,35 @@
+# frozen_string_literal: true
```
Missing require statements. This file uses JWT and SecureRandom but requires neither:

```
require 'jwt'
require 'securerandom'
```

arpit-jn commented on lib/omniauth/auth0/jwt_token.rb:
```diff
@@ -0,0 +1,35 @@
+    class JWTToken
+      attr_reader :client_id, :domain_url, :client_assertion_signing_key, :client_assertion_signing_algorithm
+
+      def initialize(client_id:, domain_url:, client_assertion_signing_key:, client_assertion_signing_algorithm: nil)
```
compatibility: The value-omitted hash shorthand (client_id: instead of client_id: client_id) requires Ruby ≥ 3.1. The gemspec declares no minimum Ruby version and the existing codebase doesn't use this syntax. This will raise a SyntaxError for users on Ruby < 3.1.

Please use explicit key-value pairs throughout, applies here and everywhere this pattern appears in the PR (specs included).

arpit-jn commented on lib/omniauth/auth0/jwt_token.rb:
```diff
@@ -0,0 +1,35 @@
+        {
+          iss: client_id,
+          sub: client_id,
+          aud: File.join(domain_url, '/'),
```
suggestion: Per [RFC 7523 #3](https://datatracker.ietf.org/doc/html/rfc7523#section-3), the aud claim should identify the authorization server, typically the token endpoint URL, not just the domain root.


`aud: File.join(domain_url, '/oauth/token')`

Auth0 may accept both, but the token endpoint URL is more spec-compliant and less ambiguous.

arpit-jn commented on lib/omniauth/strategies/auth0.rb:
```diff
@@ -4,6 +4,7 @@
 require 'uri'
 require 'securerandom'
 require 'omniauth-oauth2'
+require 'omniauth/auth0/jwt_token'
```
jwt_token.rb calls JWT.encode at runtime, but the jwt gem is only a dev/test dependency in the Gemfile, it's not in the gemspec's add_runtime_dependency. Users installing via gem install omniauth-auth0 won't have it.

Add to omniauth-auth0.gemspec:

`s.add_runtime_dependency 'jwt', '~> 2'`

(This is technically a pre-existing issue since the JWT validator also uses jwt at runtime, but it becomes critical now that JWT signing happens in the auth flow.)

arpit-jn commented on lib/omniauth/strategies/auth0.rb:
```diff
@@ -100,25 +105,20 @@ def authorize_params
       end
 
       def build_access_token
+        options.token_params.merge!(client_assertion_signing_key_token_params) if client_assertion_signing_key_auth?
```
suggestion: client_id is not included in the token request body. With the non-standard auth_scheme, the oauth2 gem won't inject it automatically. While the JWT assertion contains sub: client_id, [RFC 7523 #2.2](https://datatracker.ietf.org/doc/html/rfc7523#section-2.2) recommends also including client_id as a request parameter.

arpit-jn commented on lib/omniauth/strategies/auth0.rb:
```diff
@@ -128,10 +128,32 @@ def callback_phase
+          grant_type: AUTHORIZATION_CODE_GRANT_TYPE,
+          client_assertion_type: CLIENT_ASSERTION_TYPE,
+          client_assertion: jwt_token,
+          audience: domain_url
```
`audience: domain_url` should be removed. The audience param in the `/oauth/token` request body tells Auth0 which API to mint an access token for. It is not the same as the JWT assertion's aud claim (which identifies the authorization server).

Sending `audience: domain_url` here will request a Management API token, which will break most users' flows. If users need a specific audience, they can configure it via authorize_params or token_params.

Suggested:

```
def client_assertion_signing_key_token_params
  {
    grant_type: AUTHORIZATION_CODE_GRANT_TYPE,
    client_id: options.client_id,
    client_assertion_type: CLIENT_ASSERTION_TYPE,
    client_assertion: jwt_token
  }
end
```

arpit-jn commented on lib/omniauth/strategies/auth0.rb:
```diff
@@ -128,10 +128,32 @@ def callback_phase
 
+      def jwt_token
+        OmniAuth::Auth0::JWTToken.new(client_id: options.client_id,
+                                      domain_url:,
```
compatibility: Same Ruby 3.1+ shorthand issue. domain_url: needs to be domain_url: domain_url.

arpit-jn commented on lib/omniauth/strategies/auth0.rb:
```diff
@@ -154,20 +176,32 @@ def no_client_id?
+      def setup_client_options_auth_scheme
+        return unless client_assertion_signing_key_auth?
+
+        options.client_options.auth_scheme = :private_key_jwt
```
`private_key_jwt` is not a recognized `auth_scheme` in the oauth2 gem 1.x (which is what `omniauth-oauth2 ~> 1` pulls in). The gem only supports `:request_body` and `:basic_auth`. An unrecognized value causes the gem to silently skip all client credential injection.

This accidentally works because the JWT assertion carries the client identity, but it's fragile and relies on undocumented behavior.

Suggested fix: 
use `:request_body` and explicitly include `client_id` in `client_assertion_signing_key_token_params` while omitting `client_secret`.

arpit-jn commented on spec/omniauth/auth0/jwt_token_spec.rb:
```diff
@@ -0,0 +1,47 @@
+  let(:domain_url) { 'https://samples.auth0.com' }
+  let(:client_assertion_signing_key) { OpenSSL::PKey::RSA.generate(2048) }
+
+  describe '#jwt_token' do
```
suggestion: Consider adding a test case for EC keys (ES256) since Auth0 supports them. The `JWTToken` class should handle them already via `JWT.encode`, but a test would confirm and document the supported algorithms.

kaczowkad replied:
Within the Auth0 UI Applications -> Credentials -> Application Authentication.  With authentication method private key JWT selected 'Add New Credential' drop down only has RS256, RS384, and PS256.  Am I looking in the wrong place for ES256?

kaczowkad replied:
Added a spec using an EC key.

arpit-jn replied:
You're looking in the right place. Auth0 currently only supports RS256, RS384, and PS256 for Private Key JWT. ES256 is not available. The EC key test is still useful for verifying the `JWTToken` class handles different key types, so feel free to keep it.

arpit-jn commented on spec/omniauth/strategies/auth0_spec.rb:
```diff
@@ -14,6 +14,8 @@
   let(:domain_url) { 'https://samples.auth0.com' }
+  let(:client_assertion_signing_algorithm) { 'RS256' }
+  let(:client_assertion_signing_key) { OpenSSL::PKey::RSA.new(2048) }
   let(:application) do
```
nit: The existing codebase uses `OpenSSL::PKey::RSA.generate(2048)` (see `jwt_token_spec.rb` line 10). `.new(2048)` works but is deprecated in newer OpenSSL versions. Use `.generate` for consistency.

arpit-jn commented on spec/omniauth/strategies/auth0_spec.rb:
```diff
@@ -27,176 +29,389 @@
 
-    it 'should have the correct client secret' do
-      expect(subject[:client_secret]).to eq(client_secret)
+      it 'should have the correct client secret' do
```
nit: Nearly all the authorize-redirect tests are identical between the two auth methods, only the setup differs. Consider extracting a `shared_examples` `'authorize redirect behavior'` and including it in both contexts. Would cut ~150 lines and make these easier to maintain.

Same applies to the callback tests, `shared_examples` with a parameterized before block would reduce the duplication significantly.

arpit-jn commented on spec/omniauth/strategies/auth0_spec.rb:
```diff
@@ -290,99 +476,307 @@ def trigger_callback
+          end
+
+          it 'has basic values' do
+            expect(subject['provider']).to eq('auth0')
```
question: This stubs the token request expecting `'client_assertion' => jwt_token` which is a fixed value (`JWT.encode({ sub: client_id }, key, 'RS256'))`. But the real `JWTToken` class generates a fresh JWT with iat, exp, and jti each time. so this would never match in a real request. The test only passes because `stub_jwt_token` mocks `JWTToken.new` to return this fixed value.

This is fine for unit testing, but consider adding one integration-style test that lets the real `JWTToken` generate the assertion and uses `hash_including` without matching `client_assertion` exactly, to verify the full flow wires up correctly end-to-end.

arpit-jn commented:
Thanks for the contribution @kaczowkad this is a much-needed feature and the overall approach is solid. Apologies for the delay in reviewing this. I know this PR has been open for a while, and I appreciate your patience.

I've left detailed inline feedback across the files. The main themes are:

1. **Bugs**: syntax error in README example, `audience` in token params will break flows, missing `require` statements, and `auth_scheme: :private_key_jwt` not recognized by `oauth2` 1.x
2. **Compatibility**: Ruby 3.1+ hash shorthand will break older Rubies, `jwt` gem needs to be a runtime dependency
3. **Spec quality**: heavy duplication could be reduced with `shared_examples`

Also, please ensure all commits are GPG-signed. Commit `802638a` is currently unsigned. You can either:
- Amend and force-push with signing enabled (`git commit --amend -S`)
- Or squash into a new signed commit

See [GitHub's guide on signing commits](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits) if you need help setting that up.

Looking forward to the next iteration. 🙌

kaczowkad commented:
@arpit-jn Thank you for the feedback.  I will try to have the updates later this week.

kaczowkad committed:
PR Feedback:
Added jwt as a omniauth-auth0.gemspec runtime dependency.  Gemfile.lock reflects these changes.

JWTToken initialize method now takes positional arguments instead of keyword arguments.

OmniAuth::Strategies::Auth0 - send client_id in token request. Remove unnecessary audience parameter.
auth_scheme changed from unrecognized :private_key_jwt to :request_body.

Added testcase using EC key (ES256).

Refactored tests to use shared examples.

arpit-jn commented on spec/omniauth/strategies/auth0_spec.rb:
```diff
@@ -290,99 +403,225 @@ def trigger_callback
+                                    client_assertion_signing_key: client_assertion_signing_key,
+                                    client_assertion_signing_algorithm: client_assertion_signing_algorithm)
+            stub_jwt_token
+            stub_auth(oauth_response.merge(refresh_token:))
```
There's one remaining Ruby 3.1+ hash shorthand at
`stub_auth(oauth_response.merge(refresh_token:))`

which should be 
`stub_auth(oauth_response.merge(refresh_token: refresh_token))`

arpit-jn reviewed (approved):
LGTM
